### PR TITLE
[Workflow] Added Context to Workflow Event

### LINKED ISSUE #37421: Workflow initial_marking is optional and entered event on initial_marking is not dispatched
The 'initial_marking' option seems to be optional as the first defined place will be used as the initial marking if 'initial_marking' is not defined. It therefore only seems useful when the initial_marking differs from the first marking defined in places.

It is therefore also impossible to have a blank marking after interacting with the workflow, even if the initial_marking is not set as it will be set to the first state automatically.

On top of this, the 'entered' event is not dispatched when the object moves to its initial marking as there is no transition.

We found this out the hard way when using workflow->getMarking($object) on a newly instantiated object to set its initial place and subscribing to the 'entered' event on the initial place. 

Eventually we solved this as follows:

    appointment:
        type: state_machine
        audit_trail:
            enabled: true
        marking_store:
            type: 'method'
            property: 'marking'
        supports:
            - Appointment
        initial_marking: UNKNOWN
        places:
            - UNKNOWN
            - SCHEDULED
            - RESCHEDULED
        transitions:
            TO_SCHEDULED:
                from: UNKNOWN
                to: SCHEDULED
            TO_RESCHEDULE:
                from: 
                    - SCHEDULED
                    - RESCHEDULED
                to: RESCHEDULED

    // new object with blank state
    $appointment = new Appointment();
    $workflow = $this->workflowRegistry->get($appointment);
    // set the marking to the initial state and transition to the correct state
    $workflow->apply($appointment, 'TO_SCHEDULED');

    public static function getSubscribedEvents(): array
    {
        return [
            'workflow.patient_appointment.entered.SCHEDULED' => 'appointmentScheduled',
            'workflow.patient_appointment.entered.RESCHEDULED' => 'appointmentRescheduled'
        ];
    }

My recommendations on how to update the documentation:
- explicitly mention that 'initial_marking' is optional and that the first defined place will be used as the initial place by default.
- explicitly mention that when entering the initial place, there is no 'entered' event dispatched
- perhaps have an example of how the setup the states when the initial state should have an event.

### PR DESCRIPTION
**Pull request by epitre**
There's also a default context for the initial marking event.

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #37421 
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/13947

This time, I need context in the events, because, sometimes, the transition is automatic and sometimes, it is manual. I want to be able to make the difference. I figured using the context for that is a good idea. I hope you do too :)

By the way, I believe it also fixes #37421 . I took @pluk77 request to be able to differentiate an initial marking by having a default context in that case

### CODE REVIEW

File: src/Symfony/Bridge/Twig/Extension/WorkflowExtension.php [hunk 1]
```diff
@@ -21,6 +21,7 @@
  * WorkflowExtension.
  *
  * @author Grégoire Pineau <lyrixx@lyrixx.info>
+ * @author Carlos Pereira De Amorim <carlos@shauri.fr>
```
Comment (epitre) [inline src/Symfony/Bridge/Twig/Extension/WorkflowExtension.php:24]: I added this because I made a modification in a previous request and I forgot to add it.
Sorry

File: src/Symfony/Component/Workflow/Event/Event.php [hunk 2]
```diff
@@ -19,20 +19,23 @@
...
 class Event extends BaseEvent
 {
     private $subject;
     private $marking;
     private $transition;
     private $workflow;
+    private $context;
 
-    public function __construct(object $subject, Marking $marking, Transition $transition = null, WorkflowInterface $workflow = null)
+    public function __construct(object $subject, Marking $marking, Transition $transition = null, WorkflowInterface $workflow = null, ?array $context = null)
```
Comment (ro0NL) [inline src/Symfony/Component/Workflow/Event/Event.php:32]: is context nullability desired? an empty array would work also isnt it, giving us a consistent return type for getContext()

Comment (epitre) [reply src/Symfony/Component/Workflow/Event/Event.php:32]: this way, I make sure it is backward compatible.
I don't want to break any code already using it

Comment (ro0NL) [reply src/Symfony/Component/Workflow/Event/Event.php:32]: but `?array $context = null` vs `array $context = []` makes no difference, right?

**Comment by epitre** [comment]
appveyor is not happy but I believe it has nothing to do with me. Does it ?

File: src/Symfony/Component/Workflow/Event/Event.php (hunk 2 above)
Comment (epitre) [reply src/Symfony/Component/Workflow/Event/Event.php]: you're absolutely right!

I changed it.

Thanks !

File: src/Symfony/Component/Workflow/WorkflowInterface.php [hunk 3]
```diff
@@ -55,7 +55,9 @@ public function apply(object $subject, string $transitionName, array $context =
      *
      * @return Transition[] All enabled transitions
      */
-    public function getEnabledTransitions(object $subject);
+    public function getEnabledTransitions(object $subject, array $context = []);
+
+    public function getEnabledTransition(object $subject, string $name): ?Transition;
```
Comment (epitre) [inline src/Symfony/Component/Workflow/WorkflowInterface.php:60]: it also comes from a previous PR.

File: src/Symfony/Component/Workflow/Workflow.php [hunk 4]
```diff
@@ -30,10 +30,12 @@
  * @author Fabien Potencier <fabien@symfony.com>
  * @author Grégoire Pineau <lyrixx@lyrixx.info>
  * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ * @author Carlos Pereira De Amorim <carlos@shauri.fr>
  */
 class Workflow implements WorkflowInterface
 {
     public const DISABLE_ANNOUNCE_EVENT = 'workflow_disable_announce_event';
+    public const DEFAULT_INITIAL_CONTEXT = ['initial' => true];
```
Comment (pluk77) [inline src/Symfony/Component/Workflow/Workflow.php:38]: Is there a reason you made this an associative array and not an indexed one? 
`public const DEFAULT_INITIAL_CONTEXT = ['initial'];`

I see in one of your tests you did use an indexed array as $context.

Are yuo

Comment (epitre) [reply src/Symfony/Component/Workflow/Workflow.php:38]: this way, you can do :
` if ($context['initial']??false) `

File: src/Symfony/Component/Workflow/WorkflowInterface.php (hunk 3 above)
Comment (lyrixx) [reply src/Symfony/Component/Workflow/WorkflowInterface.php:60]: This is a BC break

File: src/Symfony/Component/Workflow/Workflow.php [hunk 5]
```diff
@@ -71,7 +73,11 @@ public function getMarking(object $subject)
             // update the subject with the new marking
             $this->markingStore->setMarking($subject, $marking);
 
-            $this->entered($subject, null, $marking);
+            if (empty($context)) {
```
Comment (lyrixx) [inline src/Symfony/Component/Workflow/Workflow.php:76]:
```suggestion
            if (!$context) {
```

But I'm not sure about this if. This is very opinionated. It could be added everytime, what do you think?

File: src/Symfony/Component/Workflow/Workflow.php [hunk 6]
```diff
@@ -309,13 +315,13 @@ private function buildTransitionBlockerListForTransition(object $subject, Markin
...
 
-    private function guardTransition(object $subject, Marking $marking, Transition $transition): ?GuardEvent
+    private function guardTransition(object $subject, Marking $marking, Transition $transition, array $context = []): ?GuardEvent
     {
         if (null === $this->dispatcher) {
             return null;
         }
 
-        $event = new GuardEvent($subject, $marking, $transition, $this);
+        $event = new GuardEvent($subject, $marking, $transition, $this, $context);
```
Comment (lyrixx) [inline src/Symfony/Component/Workflow/Workflow.php:324]: For the guard event, I always refused to pass the workflow. Because people will use it to determine if the transition is enabled or not and this is wrong

**Review by lyrixx — COMMENTED** [review]
Thanks for your PR but I already rejected in the past

File: src/Symfony/Component/Workflow/WorkflowInterface.php (hunk 3 above)
Comment (epitre) [reply src/Symfony/Component/Workflow/WorkflowInterface.php:60]: yes, of course. I'll remove it

File: src/Symfony/Component/Workflow/Workflow.php (hunk 5 above)
Comment (epitre) [reply src/Symfony/Component/Workflow/Workflow.php:76]: I like the word `empty` as it is more explicit for an array. $context can't be a boolean, nor an object.
But I'll change if you want to.

File: src/Symfony/Component/Workflow/Workflow.php (hunk 6 above)
Comment (epitre) [reply src/Symfony/Component/Workflow/Workflow.php:324]: okay.
And you think they'll do the same with the context ?

**Comment by epitre** [comment]
> Thanks for your PR but I already rejected in the past

O_O Does it mean you are rejecting this one ?

What I want, is to be able to differentiate transitions that will be done automatically from others that will be done manually.
If not with the context, how do you suggest to do that ?

**Comment by epitre** [comment]
> Thanks for your PR but I already rejected in the past

I found an old PR a realized the transition event knows the context ^^

I'll use this event, then. But I really don't understand why the other events do not deserve to know the context ^^

File: src/Symfony/Component/Workflow/Workflow.php (hunk 5 above)
Comment (lyrixx) [reply src/Symfony/Component/Workflow/Workflow.php:76]: Yes please

File: src/Symfony/Component/Workflow/Workflow.php (hunk 6 above)
Comment (lyrixx) [reply src/Symfony/Component/Workflow/Workflow.php:324]: Oups, I made a typo in my comment. I meant context and not workflow. Sorry

**Comment by lyrixx** [comment]
Oups, I write the comment but I forgot to remove it :+1:

I'm against to put the context in the guard event but for other event I guess this is fine :+1: 

> O_O Does it mean you are rejecting this one ?

nope, it's OK :+1: 

Sorry for my bad comment

**Comment by epitre** [comment]
> Oups, I write the comment but I forgot to remove it +1
> 
> I'm against to put the context in the guard event but for other event I guess this is fine +1
> 
> > O_O Does it mean you are rejecting this one ?
> 
> nope, it's OK +1
> 
> Sorry for my bad comment

Great ! So, GuardEvent is not aware of the context anymore. I simplified TransitionEvent. And TransitionEvent is still the only event which can change the context.

File: src/Symfony/Component/Workflow/Event/Event.php [hunk 7]
```diff
@@ -19,20 +19,23 @@
...
  * @author Grégoire Pineau <lyrixx@lyrixx.info>
+ * @author Carlos Pereira De Amorim <carlos@shauri.fr>
  */
 class Event extends BaseEvent
 {
     private $subject;
     private $marking;
     private $transition;
     private $workflow;
+    protected $context;
```
Comment (lyrixx) [inline src/Symfony/Component/Workflow/Event/Event.php:30]: protected must be before private

File: src/Symfony/Component/Workflow/Workflow.php [hunk 8]
```diff
@@ -286,7 +292,7 @@ public function getMetadataStore(): MetadataStoreInterface
         return $this->definition->getMetadataStore();
     }
 
-    private function buildTransitionBlockerListForTransition(object $subject, Marking $marking, Transition $transition): TransitionBlockerList
+    private function buildTransitionBlockerListForTransition(object $subject, Marking $marking, Transition $transition, array $context = []): TransitionBlockerList
```
Comment (lyrixx) [inline src/Symfony/Component/Workflow/Workflow.php:295]: $context is dead code here + it's the same for many other places

File: src/Symfony/Component/Workflow/Workflow.php [hunk 9]
```diff
@@ -395,34 +400,38 @@ private function entered(object $subject, ?Transition $transition, Marking $mark
             foreach ($transition->getTos() as $place) {
                 $this->dispatcher->dispatch($event, sprintf('workflow.%s.entered.%s', $this->name, $place));
             }
+        } elseif (!empty($this->definition->getInitialPlaces())) {
```
Comment (lyrixx) [inline src/Symfony/Component/Workflow/Workflow.php:403]: This is not related to the current PR. Could you revert it? thanks

Comment (epitre) [reply src/Symfony/Component/Workflow/Workflow.php:403]: without that, the DEFAULT_INITIAL_CONTEXT makes no sense.

it's a little far fetched, but I believe adding this context enhancement is a good moment to add this default context

Unless you don't want a DEFAULT_INITIAL_CONTEXT ?

File: src/Symfony/Component/Workflow/Event/Event.php (hunk 7 above)
Comment (epitre) [reply src/Symfony/Component/Workflow/Event/Event.php]: :+1:

File: src/Symfony/Component/Workflow/Workflow.php (hunk 8 above)
Comment (epitre) [reply src/Symfony/Component/Workflow/Workflow.php]: :+1:

**Comment by epitre** [comment]
@lyrixx up !

**Review by lyrixx — APPROVED** [review]

**Comment by lyrixx** [comment]
I'm gonna merge your PR, but I also gonna tweak 2/3 things

Thanks for your work and your patience

**Comment by lyrixx** [comment]
Thank you @epitre.

**Comment by epitre** [comment]
@lyrixx Thank you :)

Could you link me the changes you make, just to be informed, please ?

**Comment by lyrixx** [comment]
Just above: https://github.com/symfony/symfony/pull/37813

**Comment by philepsybo** [comment]
> Oups, I write the comment but I forgot to remove it 👍
> 
> I'm against to put the context in the guard event but for other event I guess this is fine 👍
> 
> > O_O Does it mean you are rejecting this one ?
> 
> nope, it's OK 👍
> 
> Sorry for my bad comment

I know this is a suuuper old issue but i would be interested to know the reasoning why the context should not be included in `GuardEvents`. If a transition can be requested by a user that is hidden behind some permission while the system should always be able to apply it if triggered by internal logic, it feels quite unintuitive to do this via `TransitionEvents` (at least for me). If i have a guard that prevents users from applying transitions they are not permitted to apply, i'd find it intuitive to have the option to convey some flag the guard so it knows "check, this is Dev explicitly telling me that this is an internal thing and i don't have to be afraid about the logged in user".
Maybe I am misinterpreting the intended semantics of `GuardEvents` and this would be an abuse of the concept.

**Comment by lyrixx** [comment]
I think a subject's transition should only be blocked via its own properties or behaviours. Not by something related to the context. This avoids scattering bits of data between the subject and the context.


Therefore, you should attach what you need to the subject itself.
